quickfixj-core/src/main/java/quickfix/ApplicationFunctionalAdapter.java

```diff
@@ -0,0 +1,301 @@
+     */
+    public <T extends Message> void addFromAppListener(Class<T> clazz, FromAppListener<T> fromAppListener) {
+        getList(fromAppTypeSafeListeners, clazz)
+                .add(fromAppListener);
+    }
+
+    /**
+     * Remove a listener of fromApp operation.
+     *
+     * @param fromAppListener the listener of fromApp operation.
+     */
+    public <T extends Message> void removeFromAppListener(FromAppListener<T> fromAppListener) {
+        fromAppListeners.remove(fromAppListener);
+        fromAppTypeSafeListeners
+                .values()
+                .forEach(list -> list.remove(fromAppListener));
+    }
+
+    @Override
+    public void onCreate(SessionID sessionId) {
+        onCreateListeners.forEach(c -> c.accept(sessionId));
+    }
+
+    @Override
+    public void onLogon(SessionID sessionId) {
+        onLogonListeners.forEach(c -> c.accept(sessionId));
+    }
+
+    @Override
+    public void onLogout(SessionID sessionId) {
+        onLogoutListeners.forEach(c -> c.accept(sessionId));
+    }
+
+    @Override
+    public void toAdmin(Message message, SessionID sessionId) {
+        toAdminListeners.forEach(c -> c.accept(message, sessionId));
+        getList(toAdminTypeSafeListeners, message.getClass())
+                .forEach(c -> c.accept(message, sessionId));
+    }
+
+    @Override
+    public void fromAdmin(Message message, SessionID sessionId) throws FieldNotFound, IncorrectDataFormat, IncorrectTagValue, RejectLogon {
+        for (FromAdminListener<Message> listener : fromAdminListeners) {
+            listener.accept(message, sessionId);
+        }
+
+        for (FromAdminListener listener : getList(fromAdminTypeSafeListeners, message.getClass())) {
+            listener.accept(message, sessionId);
+        }
+
+    }
+
+    @Override
+    public void toApp(Message message, SessionID sessionId) throws DoNotSend {
+        for (ToAppListener<Message> listener : toAppListeners) {
+            listener.accept(message, sessionId);
+        }
+
+        for (ToAppListener listener : getList(toAppTypeSafeListeners, message.getClass())) {
+            listener.accept(message, sessionId);
+        }
+    }
+
+    @Override
+    public void fromApp(Message message, SessionID sessionId) throws FieldNotFound, IncorrectDataFormat, IncorrectTagValue, UnsupportedMessageType {
+        for (FromAppListener<Message> listener : fromAppListeners) {
+            listener.accept(message, sessionId);
+        }
+
+        for (FromAppListener listener : getList(fromAppTypeSafeListeners, message.getClass())) {
+            listener.accept(message, sessionId);
+        }
+    }
+
+    private <T> List<T> getList(Map<Class, List<T>> multimap, Class clazz) {
+        List<T> list = multimap.get(clazz);
+
+        if (list != null) {
+            // Return without synchronization if the list was found
+            return list;
+        }
+
+        synchronized (multimap) {
```

Comment:
Hi @jasition , @colinduplantis 

please excuse me if I misunderstood the intention, but why don't we just use a `ConcurrentMap` and utilize `putIfAbsent()`? That way we do not need explicit synchronization and perform an atomic update.

Thanks,
Chris.

Reply:
Hi @chrjohn, 

I completely overlooked the `computeIfAbsent()` method which will eliminate the while loop to put the list to ConcurrentMap in previous commit. 

I personally would prefer `ConcurrentMap` because it is atomic. With the use of `computeIfAbsent()`, I can make an one-liner like this in the `getList` method:

`return multimap.computeIfAbsent(k -> new CopyOnWriteArrayList())`

and that should also address @colinduplantis 's concern on potential infinite loop.